### PR TITLE
Carousel calculation fixes (#561)

### LINKED ISSUE #561: Pager paddings are incorrect if there is only one page
screen shot 2016-12-01 at 15 29 19
screen shot 2016-12-02 at 15 45 07

### PR DESCRIPTION
**Pull request by alwx**
Fixes #561

### CODE REVIEW

**Comment by rasom** [comment]
The right indent of the last item in the carousel is a bit too big. I'm not sure that this is a big problem, but it looked better 

Before:

img_0737

After:

img_0739

**Comment by alwx** [comment]
Here is how it looks in Zeplin: https://app.zeplin.io/project.html#pid=570c07ce7d73d11c3c6709ae&sid=576952f6bd58e3c01fb5f9d7

So I've get all padding values there.

**Comment by rasom** [comment]
This padding for the last item is not specified in Zeplin, I'm just saying that it looks off (at least for me). So maybe we need some input from @andmironov , because it is not obvious how the nth and the last items should look in Discover carousel and in Transactions list

**Comment by alwx** [comment]
@rasom the code has been updated

**Comment by rasom** [comment]
It looks much better now. 

But still, I was able to do this:

1. When I'm swapping quickly from the last item to the first the carousel could stop on "-1th" item
2. If I touch any item at transition, animation stops and the carousel stays in invalid state

st

**Comment by alwx** [comment]
@rasom good catch. Both problems existed before and the bad news is that the second problem cannot be fixed completely because we use a ScrollView and this ScrollView responds to touch events. We can disable scrolling for some time, but we don't know for how much time (.scrollTo method doesn't have any callbacks). I've made this situation less possible by introducing a small delay, but it is still possible if you really want to reproduce it. 

The first problem is completely fixed now.

File: src/status_im/components/carousel/carousel.cljs [hunk 1]
```diff
@@ -60,14 +69,30 @@
...
   (.-x (.-contentOffset (.-nativeEvent event))))
 
+(defn get-page-position [state page]
+  (let [page-width (get-page-width state)
+        gap (get-gap state)
+        sneak (get-sneak state)
+        count (get-count state)
+        addition (condp = page
+                   0 gap
+                   (- count 1) (- (- (* sneak 2) (* 3 gap)))
```
Comment (rasom) [inline src/status_im/components/carousel/carousel.cljs]:
```clojure
(- (- (* sneak 2) (* 3 gap)))
;; is equal to
(- (* 3 gap) (* sneak 2))

(- 0 (- sneak (* gap 2)))
;; is equal to
(- (* gap 2) sneak)
```

Comment (rasom) [inline src/status_im/components/carousel/carousel.cljs]:
```clojure
(- count 1) => (dec count)
```

File: src/status_im/components/carousel/carousel.cljs [hunk 2]
```diff
@@ -60,14 +69,30 @@
...
+       (* (- page 1) gap)
+       addition)))
+
 (defn go-to-page [component page]
   (let [props (reagent.core/props component)
         state (reagent.core/state component)
         page-width (get-page-width state)
         gap (get-gap state)
-        page-position (+ (* page page-width) (* (- page 1) gap))]
+        sneak (get-sneak state)
```
Comment (rasom) [inline src/status_im/components/carousel/carousel.cljs]: `sneak` is never used

File: src/status_im/components/carousel/styles.cljs [hunk 3]
```diff
@@ -5,10 +5,15 @@
 
 (defn content-container [sneak gap]
   {:paddingLeft (+ 0 (quot gap 2))
-   :paddingRight (+ sneak (quot gap 2))})
+   :paddingRight (+ (quot gap 2))})
```
Comment (rasom) [inline src/status_im/components/carousel/styles.cljs]: it could be just `(quot gap 2)`

File: src/status_im/components/carousel/styles.cljs [hunk 4]
```diff
@@ -5,10 +5,15 @@
...
 
-(defn page [page-width margin]
-  {:width          page-width
-   :justifyContent :center
-   :marginLeft     margin
-   :marginRight    margin})
+(defn page [index count page-width gap]
+  (let [margin        (quot gap 2)
+        left-spacing  (if (= index 0) gap 0)
```
Comment (rasom) [inline src/status_im/components/carousel/styles.cljs]: `(zero? index)`

**Comment by alwx** [comment]
@rasom done

File: src/status_im/components/carousel/carousel.cljs [hunk 5]
```diff
@@ -60,14 +69,29 @@
...
 
+(defn get-page-position [state page]
+  (let [page-width (get-page-width state)
+        gap (get-gap state)
+        sneak (get-sneak state)
+        count (get-count state)
+        addition (condp = page
+                   0 gap
+                   (dec count) (- (* 3 gap) (* sneak 2))
+                   (- 0 (- (* gap 2) sneak)))]
```
Comment (rasom) [inline src/status_im/components/carousel/carousel.cljs]: `(- 0 (- a b)) => (- b a)`

File: src/status_im/components/carousel/carousel.cljs [hunk 6]
```diff
@@ -60,14 +69,29 @@
...
+    (+ (* page page-width)
+       (* (- page 1) gap)
```
Comment (rasom) [inline src/status_im/components/carousel/carousel.cljs]: `(dec page)`

File: src/status_im/components/carousel/styles.cljs [hunk 7]
```diff
@@ -5,10 +5,15 @@
 
 (defn content-container [sneak gap]
   {:paddingLeft (+ 0 (quot gap 2))
```
Comment (rasom) [inline src/status_im/components/carousel/styles.cljs]: The same problem as was on the next line. I know that it was written before, but it so close :)

File: src/status_im/components/carousel/styles.cljs [hunk 8]
```diff
@@ -5,10 +5,15 @@
...
 
-(defn page [page-width margin]
-  {:width          page-width
-   :justifyContent :center
-   :marginLeft     margin
-   :marginRight    margin})
+(defn page [index count page-width gap]
+  (let [margin        (quot gap 2)
+        left-spacing  (if (zero? index) gap 0)
+        right-spacing (if (and (= index (- count 1))
```
Comment (rasom) [inline src/status_im/components/carousel/styles.cljs]: `(dec count)`

**Comment by rasom** [comment]
thanks, @alwx

**Comment by alwx** [comment]
@rasom I've updated and checked everything. Thanks

**Pull request merged**